File: anderso2HW4/mm.py
import sys
import re

inputTsv = None
inputTmp = None

if len(sys.argv) == 3:
	inputTsv = open(sys.argv[1], "r")

# set up the records list, attribute/index dict, and template list
records = []
for line in inputTsv:
	records.append(re.split('\t', line))

attributes = {}
for i in range(len(records[0])):
	attributes[records[0][i].strip()] = i


# iterates through each record, matching the appropriate values 
# to the tags in the template
# as a note, by splitting along spaces and tabs, I lose the tabs
# when I put it back together. I could have done this without splitting
# but I got lazy
for i in range(1, len(records)):
	# output file creation
	out = open(records[i][attributes["ID"]] + ".txt", "w")

	#open up the template and put it in a list
	inputTmp = open(sys.argv[2], "r")
	template = []
	for line in inputTmp:
		template.append(re.split(' |\t', line))
	
	for line in template:
		for j in range(len(line)):
			if "<<" in line[j]:
				# obtain the start and end index of the tag and replace
				# the tag with the attribute in the records
				startIdx = line[j].index("<<")
				endIdx = line[j].index(">>")
				tag = line[j][startIdx+2:endIdx]
				before = line[j][0:startIdx]
				after = line[j][endIdx + 2:len(line[j])].strip()
				line[j] = before + records[i][attributes[tag]].strip() + after 

		# actual file io
		for tok in line:
			if tok != '\n':
				#print(tok + " ", end='')
				out.write(tok + " ")
		#print()
		out.write("\n")
	print(out.name + " written (" + str(i) + "/" + str(len(records) - 1) + ").")
	out.close()

print("Done.")	

File: anderso2HW4/mm.cs

using System;
using System.Collections.Generic;
using System.IO;
namespace MailMerge {
public class Program {	
	
	public static void getRecords(StreamReader input, out List<string[]> records) {
		records = new List<string[]>();
		while (!input.EndOfStream) {
			string[] toks = input.ReadLine().Split('\t');
			records.Add(toks);
		}
	}

	public static void writeRecords(List<string[]> records, string templateFileName) {
		// create a dictionary with the attribute name and index
		Dictionary<string, int> attributes = new Dictionary<string, int>();
		for (int i = 0; i < records[0].Length; i++) {
			attributes[records[0][i]] = i;
		}
		
		for (int i = 1; i < records.Count; i++) {
			// create the file
			string outfile = records[i][attributes["ID"]] + ".txt";
			StreamWriter output = new StreamWriter(outfile);
			StreamReader template = new StreamReader(templateFileName);
			
			while (!template.EndOfStream) {
				// get the template line
				//string[] toks = template.ReadLine().Split(' ');
				string[] toks = template.ReadLine().Split(new char[] {' ', '\t'});

				for (int j = 0; j < toks.Length; j++) {
					int tagBeginIdx = toks[j].IndexOf("<<");			
					if (tagBeginIdx >= 0) {
						//replace the tag with the appropriate parameter
						int tagEndIdx = toks[j].IndexOf(">>") + 1;
						string before = toks[j].Substring(0, tagBeginIdx - 0);
						string after = toks[j].Substring(tagEndIdx + 1, toks[j].Length - tagEndIdx - 1);
						string tagName = toks[j].Substring(tagBeginIdx + 2, tagEndIdx - tagBeginIdx - 3);
						
						toks[j] = before + records[i][attributes[tagName]] + after;
					}
		
				}
				foreach (string s in toks) {
					//Console.Write(s + " ");
					output.Write(s + " ");	
				}
				//Console.WriteLine();
				output.WriteLine();
			}
			output.Close();

		}	
	}

	public static void Main(string[] args) {
		StreamReader inputCsv = new StreamReader(args[0]);

		List<string[]> records;
		getRecords(inputCsv, out records);		
		writeRecords(records, args[1]);
		Console.WriteLine("Done.");
	}
}
}
